src/main/kotlin/com/terraformation/backend/accelerator/ParticipantProjectSpeciesService.kt

```diff
@@ -52,29 +53,42 @@ class ParticipantProjectSpeciesService(
     return dslContext.transactionResult { _ ->
       val existingModels = participantProjectSpeciesStore.create(projectIds, speciesIds)
 
-      val checkedProjectIds = emptySet<ProjectId>()
+      // Used to save relatively expensive queries for projects which we know have submissions
+      val projectDeliverableIds = mutableMapOf<ProjectId, DeliverableId>()
 
-      existingModels.forEach {
-        // A submission must exist for every project that is getting a new species assigned
-        if (checkedProjectIds.contains(it.projectId)) {
+      existingModels.forEach { participantProjectSpecies ->
+        if (projectDeliverableIds.contains(participantProjectSpecies.projectId)) {
+          publishEditEvent(
+              projectDeliverableIds[participantProjectSpecies.projectId]!!,
+              participantProjectSpecies)
           return@forEach
         }
 
+        // A submission must exist for every project that is getting a new species assigned
         val deliverableSubmission =
-            submissionStore.fetchActiveSpeciesDeliverableSubmission(it.projectId)
+            submissionStore.fetchActiveSpeciesDeliverableSubmission(
+                participantProjectSpecies.projectId)
         if (deliverableSubmission.submissionId == null) {
-          submissionStore.createSubmission(deliverableSubmission.deliverableId, it.projectId)
+          submissionStore.createSubmission(
+              deliverableSubmission.deliverableId, participantProjectSpecies.projectId)
         }
 
-        checkedProjectIds.plus(it.projectId)
-
-        eventPublisher.publishEvent(
-            ParticipantProjectSpeciesAddedEvent(
-                deliverableId = deliverableSubmission.deliverableId,
-                participantProjectSpecies = it))
+        projectDeliverableIds[participantProjectSpecies.projectId] =
+            deliverableSubmission.deliverableId
+        publishEditEvent(
+            projectDeliverableIds[participantProjectSpecies.projectId]!!, participantProjectSpecies)
       }
 
       existingModels
     }
   }
+
+  private fun publishEditEvent(
```

Comment:
Maybe call this `publishAddedEvent` since there is also an "edited" event which people might assume this publishes.